apollo-router-core/src/layers/deduplication.rs

```diff
@@ -0,0 +1,131 @@
+use tower::{BoxError, Layer, ServiceExt};
+
+pub struct QueryDeduplicationLayer;
+
+impl<S> Layer<S> for QueryDeduplicationLayer
+where
+    S: tower::Service<SubgraphRequest, Response = SubgraphResponse, Error = BoxError> + Clone,
+{
+    type Service = QueryDeduplicationService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        QueryDeduplicationService::new(service)
+    }
+}
+
+pub struct QueryDeduplicationService<S> {
+    service: S,
+    wait_map: Arc<
+        Mutex<HashMap<http_compat::Request<Request>, Sender<Result<SubgraphResponse, String>>>>,
+    >,
+}
+
+impl<S> QueryDeduplicationService<S>
+where
+    S: tower::Service<SubgraphRequest, Response = SubgraphResponse, Error = BoxError> + Clone,
+{
+    fn new(service: S) -> Self {
+        QueryDeduplicationService {
+            service,
+            wait_map: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    async fn dedup(
+        service: S,
+        wait_map: Arc<
+            Mutex<HashMap<http_compat::Request<Request>, Sender<Result<SubgraphResponse, String>>>>,
+        >,
+        request: SubgraphRequest,
+    ) -> Result<SubgraphResponse, BoxError> {
+        loop {
+            let mut locked_wait_map = wait_map.lock().await;
+            match locked_wait_map.get_mut(&request.http_request) {
+                Some(waiter) => {
+                    // Register interest in key
+                    let mut receiver = waiter.subscribe();
+                    drop(locked_wait_map);
+
+                    match receiver.recv().await {
+                        Ok(value) => {
+                            return value
+                                .map(|response| SubgraphResponse {
+                                    response: response.response,
+                                    context: request.context,
+                                })
+                                .map_err(|e| e.into())
+                        }
+                        // there was an issue with the broadcast channel, retry fetching
+                        Err(_) => continue,
+                    }
+                }
+                None => {
+                    let (tx, _rx) = broadcast::channel(1);
+                    locked_wait_map.insert(request.http_request.clone(), tx.clone());
+                    drop(locked_wait_map);
+
+                    let context = request.context.clone();
+                    let http_request = request.http_request.clone();
+                    let res = service.ready_oneshot().await?.call(request).await;
+
+                    {
+                        let mut locked_wait_map = wait_map.lock().await;
+                        locked_wait_map.remove(&http_request);
+                    }
+
+                    // Let our waiters know
+                    let broadcast_value = res
+                        .as_ref()
+                        .map(|response| response.clone())
+                        .map_err(|e| e.to_string());
```

Comment:
We use `BoxError` all over the code (mainly because [Buffer enforces BoxError](https://docs.rs/tower/latest/tower/buffer/struct.Buffer.html#impl)), and `BoxError` is not `Clone`. But here we want to copy the response to all of the waiting queries, even if we got an error. Since we cannot access the underlying type, the best we can do is convert to a String and pass it around

Reply:
we could serialize a JSON value to a string and pass it as error. Unfortunately, the only way to interact with `std::error::Error` is through strings: https://doc.rust-lang.org/nightly/std/error/trait.Error.html

Reply:
IIRC we discussed (with @BrynCooke I think?) using `FetchError` as the error type for all subgraph services and layers, and convert to `BoxError` just before passing it back to the `Buffer` layer, but deferred that to later as that would be a big change introduced by that PR